File: packages/js/cli/src/commands/publish.ts
/**
 * trik publish command
 *
 * Publishes a trik to the TrikHub registry.
 *
 * New simplified flow:
 * 1. Validate trik structure using @trikhub/linter
 * 2. Verify git tag exists on remote
 * 3. Get commit SHA for the tag
 * 4. Register with TrikHub registry (no tarball, no GitHub Release)
 *
 * Supports both Node.js and Python package structures:
 * - Node.js: manifest.json at repository root, dist/ must be committed
 * - Python: manifest.json inside package subdirectory (e.g., package_name/manifest.json)
 */

import { existsSync, readFileSync } from 'node:fs';
import { join, resolve, relative } from 'node:path';
import { execSync } from 'node:child_process';
import chalk from 'chalk';
import ora from 'ora';
import { TrikLinter, findManifestPath } from '@trikhub/linter';
import { TrikHubMetadata } from '../types.js';
import { RegistryClient } from '../lib/registry.js';
import { loadConfig } from '../lib/storage.js';

interface PublishOptions {
  directory?: string;
  tag?: string;
}

interface TrikManifest {
  id: string;
  name: string;
  version: string;
  description: string;
  entry: {
    module: string;
    export: string;
    runtime?: 'node' | 'python';
  };
  actions: Record<string, unknown>;
  capabilities?: unknown;
  limits?: unknown;
}

/**
 * Get the commit SHA that a tag points to on the remote
 */
function getRemoteTagCommitSha(trikDir: string, tag: string): string | null {
  try {
    // First try to get the tag's commit from the remote
    const result = execSync(`git ls-remote --tags origin refs/tags/${tag}`, {
      cwd: trikDir,
      encoding: 'utf-8',
      stdio: ['pipe', 'pipe', 'pipe'],
    }).trim();

    if (!result) {
      return null;
    }

    // Format is: "sha\trefs/tags/tagname"
    const sha = result.split('\t')[0];

    // If it's an annotated tag, we need to dereference it to get the commit SHA
    const derefResult = execSync(`git ls-remote --tags origin refs/tags/${tag}^{}`, {
      cwd: trikDir,
      encoding: 'utf-8',
      stdio: ['pipe', 'pipe', 'pipe'],
    }).trim();

    if (derefResult) {
      // Annotated tag - use dereferenced commit
      return derefResult.split('\t')[0];
    }

    // Lightweight tag - use the SHA directly
    return sha;
  } catch {
    return null;
  }
}

/**
 * Check if the dist/ directory is committed (not ignored)
 */
function isDistCommitted(repoDir: string): boolean {
  try {
    // Check if dist/ is tracked in git
    const result = execSync('git ls-files dist/', {
      cwd: repoDir,
      encoding: 'utf-8',
      stdio: ['pipe', 'pipe', 'pipe'],
    }).trim();

    return result.length > 0;
  } catch {
    return false;
  }
}

/**
 * Check if a file/directory is committed (not ignored)
 */
function isPathCommitted(repoDir: string, relativePath: string): boolean {
  try {
    const result = execSync(`git ls-files "${relativePath}"`, {
      cwd: repoDir,
      encoding: 'utf-8',
      stdio: ['pipe', 'pipe', 'pipe'],
    }).trim();

    return result.length > 0;
  } catch {
    return false;
  }
}

export async function publishCommand(options: PublishOptions): Promise<void> {
  const spinner = ora();
  const config = loadConfig();

  // Check if logged in
  if (!config.authToken) {
    console.log(chalk.red('Not logged in'));
    console.log(chalk.dim('Run `trik login` to authenticate first'));
    process.exit(1);
  }

  // Check if token is expired
  if (config.authExpiresAt && new Date(config.authExpiresAt) < new Date()) {
    console.log(chalk.red('Session expired'));
    console.log(chalk.dim('Run `trik login` to re-authenticate'));
    process.exit(1);
  }

  const repoDir = resolve(options.directory || '.');

  try {
    // Step 1: Validate trik structure using the centralized linter
    spinner.start('Validating trik structure...');

    // Find manifest.json (supports both Node.js and Python package structures)
    const manifestLocation = await findManifestPath(repoDir);

    if (!manifestLocation) {
      spinner.fail('Missing manifest.json');
      console.log(chalk.dim('Create a manifest.json file with your trik definition'));
      console.log(chalk.dim('\nFor Node.js triks: place manifest.json at repository root'));
      console.log(chalk.dim('For Python triks: place manifest.json inside your package directory'));
      process.exit(1);
    }

    const { manifestPath, manifestDir, packageType } = manifestLocation;

    // Log detected package type
    if (packageType === 'python') {
      const relPath = relative(repoDir, manifestDir);
      console.log(chalk.dim(`  Detected Python package: ${relPath}/`));
    }

    // Check trikhub.json exists (always at repo root)
    const trikhubPath = join(repoDir, 'trikhub.json');
    if (!existsSync(trikhubPath)) {
      spinner.fail('Missing trikhub.json');
      console.log(chalk.dim('Create a trikhub.json file with registry metadata'));
      process.exit(1);
    }

    // Read manifest for later use (version, id, etc.)
    let manifest: TrikManifest;
    try {
      manifest = JSON.parse(readFileSync(manifestPath, 'utf-8'));
    } catch (error) {
      spinner.fail('Invalid manifest.json');
      console.log(chalk.red(error instanceof Error ? error.message : 'Parse error'));
      process.exit(1);
    }

    // Read trikhub.json
    let trikhubMeta: TrikHubMetadata;
    try {
      trikhubMeta = JSON.parse(readFileSync(trikhubPath, 'utf-8'));
    } catch (error) {
      spinner.fail('Invalid trikhub.json');
      console.log(chalk.red(error instanceof Error ? error.message : 'Parse error'));
      process.exit(1);
    }

    // Compute entry path for later git commit check
    const entryPath = join(manifestDir, manifest.entry.module);

    // Run validation using the centralized linter (includes entry point check)
    const linter = new TrikLinter({
      checkCompiledEntry: true, // Verify compiled entry point exists
      skipRules: ['manifest-completeness'], // Skip optional field warnings for publish
    });
    const results = await linter.lintManifestOnly(repoDir);
    const hasErrors = linter.hasErrors(results);
    const warnings = results.filter((r) => r.severity === 'warning');

    if (hasErrors) {
      spinner.fail('Validation failed');
      console.log(linter.formatResults(results));
      process.exit(1);
    }

    if (warnings.length > 0) {
      spinner.warn('Validation passed with warnings');
      console.log(linter.formatResults(results));
    } else {
      spinner.succeed('Validation passed');
    }

    // Step 2: Determine version and git tag
    const version = options.tag?.replace(/^v/, '') || manifest.version;
    const gitTag = `v${version}`;
    console.log(chalk.dim(`  Version: ${version}`));

    // Step 3: Get GitHub repo from trikhub.json
    const repoUrl = trikhubMeta.repository;
    const repoMatch = repoUrl.match(/github\.com\/([^/]+\/[^/]+)/);
    if (!repoMatch) {
      console.log(chalk.red('Invalid repository URL in trikhub.json'));
      console.log(chalk.dim('Expected format: https://github.com/owner/repo'));
      process.exit(1);
    }
    const githubRepo = repoMatch[1].replace(/\.git$/, '');
    const [owner] = githubRepo.split('/');
    const trikName = manifest.id || manifest.name;
    const fullName = `@${owner}/${trikName}`;

    // Step 4: Verify git remote matches trikhub.json repository
    spinner.start('Verifying git remote...');
    try {
      const gitRemote = execSync('git remote get-url origin', {
        cwd: repoDir,
        encoding: 'utf-8',
        stdio: ['pipe', 'pipe', 'pipe'],
      }).trim();

      // Normalize URLs for comparison (handle SSH and HTTPS variants)
      const normalizeGitUrl = (url: string): string => {
        return url
          .replace(/^git@github\.com:/, 'github.com/')
          .replace(/^https?:\/\//, '')
          .replace(/\.git$/, '')
          .toLowerCase();
      };

      const normalizedRemote = normalizeGitUrl(gitRemote);

      if (!normalizedRemote.includes(githubRepo.toLowerCase())) {
        spinner.fail('Git remote does not match trikhub.json repository');
        console.log(chalk.red('\nRepository mismatch detected:'));
        console.log(chalk.dim(`  trikhub.json: ${repoUrl}`));
        console.log(chalk.dim(`  git remote:   ${gitRemote}`));
        console.log();
        console.log(chalk.dim('Update trikhub.json to match your git remote, or push to the correct repository.'));
        process.exit(1);
      }
      spinner.succeed('Git remote verified');
    } catch {
      // Not a git repo or no remote configured
      spinner.fail('Not a git repository or no remote configured');
      console.log(chalk.dim('Initialize git and add a remote that matches trikhub.json:'));
      console.log(chalk.dim(`  git init`));
      console.log(chalk.dim(`  git remote add origin ${repoUrl}`));
      process.exit(1);
    }

    console.log(chalk.dim(`  Trik: ${fullName}`));
    console.log(chalk.dim(`  Repo: ${githubRepo}`));

    // Step 5: Check that entry module is committed
    if (packageType === 'node') {
      // Node.js: check dist/ is committed
      spinner.start('Checking dist/ is committed...');
      if (!isDistCommitted(repoDir)) {
        spinner.fail('dist/ directory is not committed to git');
        console.log(chalk.red('\nTriks require dist/ to be committed for direct GitHub installation.'));
        console.log(chalk.dim('Add dist/ to your repository:'));
        console.log(chalk.dim(`  git add dist/ -f`));
        console.log(chalk.dim(`  git commit -m "Add dist for publishing"`));
        console.log(chalk.dim(`  git push`));
        process.exit(1);
      }
      spinner.succeed('dist/ is committed');
    } else {
      // Python: check entry module is committed
      const entryRelPath = relative(repoDir, entryPath);
      spinner.start(`Checking ${entryRelPath} is committed...`);
      if (!isPathCommitted(repoDir, entryRelPath)) {
        spinner.fail(`Entry module is not committed to git`);
        console.log(chalk.red(`\n${entryRelPath} must be committed for direct GitHub installation.`));
        console.log(chalk.dim('Add it to your repository:'));
        console.log(chalk.dim(`  git add "${entryRelPath}"`));
        console.log(chalk.dim(`  git commit -m "Add entry module for publishing"`));
        console.log(chalk.dim(`  git push`));
        process.exit(1);
      }
      spinner.succeed(`${entryRelPath} is committed`);
    }

    // Step 6: Verify git tag exists on remote
    spinner.start(`Verifying tag ${gitTag} exists on remote...`);
    const commitSha = getRemoteTagCommitSha(repoDir, gitTag);

    if (!commitSha) {
      spinner.fail(`Tag ${gitTag} not found on remote`);
      console.log(chalk.red('\nThe git tag must exist on the remote before publishing.'));
      console.log(chalk.dim('Create and push the tag:'));
      console.log(chalk.dim(`  git tag ${gitTag}`));
      console.log(chalk.dim(`  git push origin ${gitTag}`));
      process.exit(1);
    }
    spinner.succeed(`Tag ${gitTag} verified (${commitSha.slice(0, 8)}...)`);

    // Step 7: Register with registry
    spinner.start('Publishing to TrikHub registry...');

    const registry = new RegistryClient();

    try {
      // Check if trik exists, if not register it
      const existingTrik = await registry.getTrik(fullName);

      if (!existingTrik) {
        // Register new trik
        try {
          await registry.registerTrik({
            githubRepo,
            name: trikName, // Explicit name from manifest.id
            description: trikhubMeta.shortDescription || manifest.description,
            categories: trikhubMeta.categories,
            keywords: trikhubMeta.keywords,
          });
          console.log(chalk.dim(`  Registered new trik: ${fullName}`));
        } catch (regError) {
          // If trik already exists (409), that's fine - continue to publish version
          // This can happen if the trik was registered but GET didn't find it
          if (regError instanceof Error && !regError.message.includes('already exists')) {
            throw regError;
          }
          console.log(chalk.dim(`  Trik already registered: ${fullName}`));
        }
      }

      // Publish version
      await registry.publishVersion(fullName, {
        version,
        gitTag,
        commitSha,
        manifest: manifest as unknown as Record<string, unknown>,
      });

      spinner.succeed('Published to TrikHub registry');
    } catch (error) {
      spinner.fail('Failed to publish to registry');
      console.log(chalk.red(error instanceof Error ? error.message : String(error)));
      process.exit(1);
    }

    // Success message
    console.log();
    console.log(chalk.green.bold('  Published successfully!'));
    console.log();
    console.log(`  ${chalk.dim('Install with:')} trik install ${fullName}@${version}`);
    console.log(`  ${chalk.dim('View at:')} https://trikhub.com/triks/${encodeURIComponent(fullName)}`);
    console.log();
  } catch (error) {
    spinner.fail('Publish failed');
    if (error instanceof Error) {
      console.error(chalk.red(error.message));
    }
    process.exit(1);
  }
}

File: packages/python/trikhub/cli/commands/publish.py
"""trik publish / unpublish — publish triks to the registry."""

from __future__ import annotations

import asyncio
import json
import re
import subprocess
import sys
from pathlib import Path

import click

from trikhub.cli.config import is_auth_expired, read_global_config
from trikhub.cli.output import ok, fail, info, warn
from trikhub.cli.registry import get_registry
from trikhub.linter.scanner import scan_capabilities, cross_check_manifest
from trikhub.manifest import validate_manifest


def _find_manifest(repo_dir: Path) -> tuple[Path, dict] | None:
    """Find and parse manifest.json in the repo."""
    manifest_path = repo_dir / "manifest.json"
    if not manifest_path.exists():
        # Check subdirectories for v1 layout
        for child in repo_dir.iterdir():
            if child.is_dir() and not child.name.startswith((".", "_")):
                candidate = child / "manifest.json"
                if candidate.exists():
                    manifest_path = candidate
                    break
        else:
            return None

    try:
        data = json.loads(manifest_path.read_text(encoding="utf-8"))
        return manifest_path, data
    except (json.JSONDecodeError, OSError):
        return None


def _find_trikhub_json(repo_dir: Path) -> dict | None:
    """Find and parse trikhub.json."""
    path = repo_dir / "trikhub.json"
    if not path.exists():
        return None
    try:
        return json.loads(path.read_text(encoding="utf-8"))
    except (json.JSONDecodeError, OSError):
        return None


def _extract_github_repo(url: str) -> str | None:
    """Extract owner/repo from a GitHub URL, preserving case."""
    match = re.search(r"github\.com[/:]([^/]+/[^/]+)", url)
    if not match:
        return None
    return re.sub(r"\.git$", "", match.group(1))


def _normalize_git_url(url: str) -> str:
    """Normalize a git URL for comparison (lowercased)."""
    url = url.rstrip("/")
    url = re.sub(r"\.git$", "", url)
    url = re.sub(r"^https?://github\.com/", "", url)
    url = re.sub(r"^git@github\.com:", "", url)
    return url.lower()


def _get_remote_url(repo_dir: Path) -> str | None:
    """Get the git remote origin URL."""
    try:
        result = subprocess.run(
            ["git", "remote", "get-url", "origin"],
            cwd=str(repo_dir),
            capture_output=True,
            text=True,
            check=True,
        )
        return result.stdout.strip()
    except subprocess.CalledProcessError:
        return None


def _get_remote_tag_sha(repo_dir: Path, tag: str) -> str | None:
    """Get the commit SHA for a remote tag."""
    try:
        result = subprocess.run(
            ["git", "ls-remote", "--tags", "origin", tag],
            cwd=str(repo_dir),
            capture_output=True,
            text=True,
            check=True,
        )
        output = result.stdout.strip()
        if output:
            return output.split()[0]
        return None
    except subprocess.CalledProcessError:
        return None


def _is_path_committed(repo_dir: Path, rel_path: str) -> bool:
    """Check if a file path is committed to git."""
    try:
        result = subprocess.run(
            ["git", "ls-files", rel_path],
            cwd=str(repo_dir),
            capture_output=True,
            text=True,
            check=True,
        )
        return bool(result.stdout.strip())
    except subprocess.CalledProcessError:
        return False


@click.command("publish")
@click.option("-d", "--directory", default=".", help="Repository directory")
@click.option("-t", "--tag", default=None, help="Git tag override")
def publish_command(directory: str, tag: str | None) -> None:
    """Validate and publish a trik to the TrikHub registry."""
    repo_dir = Path(directory).resolve()

    # Step 1: Check auth
    global_config = read_global_config()
    if not global_config.auth_token or is_auth_expired(global_config):
        fail("Not authenticated")
        info("Run `trik login` to authenticate first")
        sys.exit(1)

    # Step 2: Find and validate manifest
    result = _find_manifest(repo_dir)
    if not result:
        fail("Missing manifest.json")
        info("Create a manifest.json file with your trik definition")
        sys.exit(1)
    manifest_path, manifest_data = result

    # Check trikhub.json
    trikhub_json = _find_trikhub_json(repo_dir)
    if not trikhub_json:
        fail("Missing trikhub.json")
        info("Create a trikhub.json file with registry metadata")
        sys.exit(1)

    # Check entry point exists
    manifest_dir = manifest_path.parent
    entry_module = manifest_data.get("entry", {}).get("module", "")
    entry_path = manifest_dir / entry_module
    if not entry_path.exists():
        fail(f"Entry point not found: {entry_module}")
        sys.exit(1)

    # Validate manifest
    validation = validate_manifest(manifest_data)
    if not validation.valid:
        fail("Validation failed")
        for error in validation.errors:
            info(f"  {error}")
        sys.exit(1)

    if validation.warnings:
        warn("Validation passed with warnings")
        for warning in validation.warnings:
            info(warning)
    else:
        ok("Validation passed")

    # Cross-check: scanner results vs manifest declarations
    manifest_dir = manifest_path.parent
    scan_result = scan_capabilities(manifest_dir)
    cross_check_errors = cross_check_manifest(scan_result, manifest_data)
    if cross_check_errors:
        fail("Capability cross-check failed")
        for err in cross_check_errors:
            location_str = ""
            if err["locations"]:
                loc = err["locations"][0]
                location_str = f" ({loc['file']}:{loc['line']})"
            info(f"  {err['message']}{location_str}")
        sys.exit(1)

    # Step 3: Determine version and git tag
    version = manifest_data.get("version", "0.1.0")
    git_tag = tag or f"v{version}"
    info(f"Version: {version}")

    # Step 4: Extract GitHub repo from trikhub.json
    repo_url = trikhub_json.get("repository", "")
    github_repo = _extract_github_repo(repo_url)

    if not github_repo:
        fail("Invalid repository URL in trikhub.json")
        info("Expected format: https://github.com/owner/repo")
        sys.exit(1)

    # Step 5: Verify git remote matches trikhub.json repository
    remote_url = _get_remote_url(repo_dir)
    if remote_url:
        normalized_remote = _normalize_git_url(remote_url)
        normalized_repo = _normalize_git_url(repo_url)
        if normalized_remote != normalized_repo:
            fail("Git remote does not match trikhub.json repository")
            info(f"trikhub.json: {repo_url}")
            info(f"git remote:   {remote_url}")
            sys.exit(1)
        ok("Git remote verified")
    else:
        fail("Not a git repository or no remote configured")
        info(f"git remote add origin {repo_url}")
        sys.exit(1)

    owner = github_repo.split("/")[0]
    trik_name = manifest_data.get("id") or manifest_data.get("name", "")
    full_name = f"@{owner}/{trik_name}"
    info(f"Trik: {full_name}")
    info(f"Repo: {github_repo}")

    # Step 6: Check entry point is committed
    entry_rel = str(manifest_path.parent.relative_to(repo_dir) / entry_module)
    if not _is_path_committed(repo_dir, entry_rel):
        fail(f"Entry module is not committed to git")
        info(f"{entry_rel} must be committed for direct GitHub installation.")
        info(f'git add "{entry_rel}"')
        info(f'git commit -m "Add entry module for publishing"')
        info("git push")
        sys.exit(1)
    ok(f"{entry_rel} is committed")

    # Step 7: Verify tag exists on remote
    commit_sha = _get_remote_tag_sha(repo_dir, git_tag)
    if not commit_sha:
        fail(f"Tag {git_tag} not found on remote")
        info("The git tag must exist on the remote before publishing.")
        info(f"git tag {git_tag}")
        info(f"git push origin {git_tag}")
        sys.exit(1)
    ok(f"Tag {git_tag} verified ({commit_sha[:8]}...)")

    # Step 8: Publish to registry
    async def _publish():
        async with get_registry() as registry:
            # Register trik if needed
            try:
                existing = await registry.get_trik(full_name)
                if not existing:
                    await registry.register_trik(
                        github_repo=github_repo,
                        name=trik_name,
                        description=manifest_data.get("description"),
                        categories=trikhub_json.get("categories"),
                        keywords=trikhub_json.get("keywords"),
                    )
                    info(f"Registered new trik: {full_name}")
            except RuntimeError:
                pass  # Already registered

            await registry.publish_version(
                full_name=full_name,
                version=version,
                git_tag=git_tag,
                commit_sha=commit_sha,
                manifest=manifest_data,
            )
            ok("Published to TrikHub registry")

    try:
        asyncio.run(_publish())
    except Exception as exc:
        fail("Failed to publish to registry")
        info(str(exc))
        sys.exit(1)

    # Success message
    click.echo()
    click.echo(click.style("  Published successfully!", fg="green", bold=True))
    click.echo()
    info(f"Install with: trik install {full_name}@{version}")
    info(f"View at: https://trikhub.com/triks/{full_name}")
    click.echo()


@click.command("unpublish")
@click.argument("package")
def unpublish_command(package: str) -> None:
    """Remove a trik from the registry."""
    global_config = read_global_config()
    if not global_config.auth_token or is_auth_expired(global_config):
        click.echo(click.style("  Not authenticated. Run `trik login` first.", fg="red"))
        sys.exit(1)

    click.echo(click.style(
        f"\n  WARNING: This will permanently delete {package} from the registry.\n",
        fg="red",
        bold=True,
    ))

    confirm = click.prompt("Type the trik name to confirm")
    if confirm != package:
        click.echo("  Aborted.")
        return

    async def _unpublish():
        async with get_registry() as registry:
            await registry.delete_trik(package)
            click.echo(click.style(f"  Unpublished {package}", fg="green"))

    asyncio.run(_unpublish())
